### PR TITLE
GPTQ support on ROCm

### PR DESCRIPTION
**Pull request by fxmarty**
Tested with
```
CUDA_VISIBLE_DEVICES=0 text-generation-launcher --model-id TheBloke/Llama-2-7B-Chat-GPTQ --quantize gptq
EXLLAMA_VERSION=1 CUDA_VISIBLE_DEVICES=0 text-generation-launcher --model-id TheBloke/Llama-2-7B-Chat-GPTQ --quantize gptq
CUDA_VISIBLE_DEVICES="0,1" text-generation-launcher --model-id TheBloke/Llama-2-7B-Chat-GPTQ --quantize gptq
```

all with good and identical results on MI210.

### CODE REVIEW

File: Dockerfile_amd [hunk 1]
```diff
@@ -75,8 +75,8 @@ RUN chmod +x ~/mambaforge.sh && \
     mamba init && \
     rm ~/mambaforge.sh
 
-# Install PyTorch nightly (2.2.0.dev2023) compiled against RoCm 5.7, as VLLM can not be compiled with RoCm 5.6.
-RUN pip install --pre torch==2.2.0.dev20231106 --index-url https://download.pytorch.org/whl/nightly/rocm5.7
```
Comment (fxmarty) [inline Dockerfile_amd:79]: This nightly version was removed from https://download.pytorch.org/whl/nightly/rocm5.7

File: server/text_generation_server/utils/gptq/exllamav2.py [hunk 2]
```diff
@@ -167,7 +166,11 @@ def __init__(self, qweight, qzeros, scales, g_idx, bias, bits, groupsize):
         self.qweight = qweight
         self.qzeros = qzeros
         self.scales = scales
-        self.g_idx = g_idx
+
+        if torch.equal(g_idx, torch.tensor([i // groupsize for i in range(self.infeatures)], dtype=torch.int32, device=g_idx.device)):
+            self.g_idx = None
+        else:
+            self.g_idx = g_idx
```
Comment (fxmarty) [inline server/text_generation_server/utils/gptq/exllamav2.py:173]: Fixes a bug in the original exllamav2 code.

File: server/text_generation_server/utils/gptq/exllamav2.py [hunk 3]
```diff
@@ -185,6 +188,11 @@ def post_init(self, temp_dq):
             "g_idx": self.g_idx,
         }
         temp_dq = temp_dq.get_scratch_slice(self.temp_dq_size())
+
+        # We NEED to keep a pointer on Python side, otherwise the garbage collector will mess with us,
+        # and `Memory access fault by GPU node-2` will EAT you.
+        self.temp_dq = temp_dq
```
Comment (fxmarty) [inline server/text_generation_server/utils/gptq/exllamav2.py:194]: This gave me headaches today.

File: Dockerfile_amd [hunk 4]
```diff
@@ -149,5 +169,5 @@ ENTRYPOINT ["./entrypoint.sh"]
 # Final image
 FROM base-copy
 
-ENTRYPOINT ["text-generation-launcher"]
-CMD ["--json-output"]
+#ENTRYPOINT ["text-generation-launcher"]
+#CMD ["--json-output"]
```
Comment (OlivierDehaene) [inline Dockerfile_amd:173]:
```suggestion
ENTRYPOINT ["text-generation-launcher"]
CMD ["--json-output"]
```

**Review by OlivierDehaene — COMMENTED** [review]
Thanks
We should really setup a CI for RocM now

File: server/text_generation_server/utils/gptq/exllamav2.py (hunk 2 above)
Comment (Narsil) [reply server/text_generation_server/utils/gptq/exllamav2.py:173]: Wow THANKS

File: server/text_generation_server/utils/gptq/exllamav2.py (hunk 3 above)
Comment (Narsil) [reply server/text_generation_server/utils/gptq/exllamav2.py:194]: THIS is what was crashing TP>1?

Comment (OlivierDehaene) [reply server/text_generation_server/utils/gptq/exllamav2.py:194]: @narsil can we test this quickly? If that's the case that would be amazing.

Comment (Narsil) [reply server/text_generation_server/utils/gptq/exllamav2.py:194]: https://github.com/huggingface/text-generation-inference/pull/1490

**Comment by OlivierDehaene** [comment]
Can you rebase + python update_doc.py?

**Pull request merged**